python/housinginsights/ingestion/LoadData.py

```diff
@@ -286,8 +289,267 @@ def rebuild(self):
+
+                # derive column and values strings needed for sql query
+                columns = ', '.join(columns)
+                columns = 'zone, ' + columns
+
+                values = ', '.join(values)
+                values = "'" + zone + "', " + values
+
+                q = "INSERT INTO zone_facts ({cols}) VALUES ({vals})".format(
+                    cols=columns, vals=values)
+
+                with self.engine.connect() as conn:
+                    result = conn.execute(q)
+                    query_results.append(result)
+
+        return query_results
+
+    def _get_zone_specifics_for_zone_type(self, zone_type):
+        """
+        Returns a list of zone_specific values for a given zone_type.
+        """
+        with self.engine.connect() as conn:
+
+            if zone_type == 'ward':
+                table = 'project'
+            elif zone_type == 'neighborhood_cluster':
+                table = 'census_tract_to_neighborhood_cluster'
+            else:
+                table = 'census'
+
+            query_result = conn.execute(
+                'select distinct {zone} from {table};'.format(zone=zone_type,
+                                                              table=table))
+            zone_specifics = [row[0] for row in query_result.fetchall()]
+            # zones.append(zone)
+
+        return zone_specifics
+
+    def _items_divide(self, numerator_data, denominator_data):
+        """
+        Divides items in the numerator by items in the denominator by matching
+        the appropriate groupings.
+
+        Takes data that is formatted for output the API, i.e. a dictionary
+        with key "items", which contains a list of dictionaries each with 'grouping'
+        and 'count'.
+
+        Returns items as dictionary with group and division result as
+        key/value pairs instead of list of dictionaries.
+        """
+        items = {}
+        if numerator_data['items'] is None:
+            items = None
+        else:
+            for n in numerator_data['items']:
+                # TODO what should we do when a matching item isn't found?
+                matching_d = next((item for item in denominator_data['items'] if
+                                   item['group'] == n['group']),
+                                  {'group': '_unknown', 'value': None})
+                if matching_d['value'] is None or n['value'] is None:
+                    divided = None
+                else:
+                    divided = n['value'] / matching_d['value']
+
+                # item = dict({'group': n['group'],
+                #              'value': divided})
+                items[n['group']] = divided
+
+        return {'items': items, 'grouping': numerator_data['grouping'],
+                'data_id': numerator_data['data_id']}
+
+    def _census_with_weighting(self, data_id, grouping):
+        """
+        Zone facts table helper function to get data from our census table.\
```

Comment:
Would be good to add to docstring: 'returns weighted data for either a) a column from the 'census' table or b) a calculated field defined by a keyword that has a custom calculation defined in this method.` or something liek that